Implement setup.cfg handling for Pip

chmeliik opened the pull request:
* CLOUDBLD-2053

Aims for near parity with the setuptools way of handling setup.cfg files
with the notable exception of never executing unknown Python modules.

There may be other minor differences, but should cover 99% of the
typical use cases.

chmeliik commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+    def _resolve_version(self, version):
+        """Attempt to resolve the version attribute."""
+        if version.startswith("file:"):
+            file_arg = version[len("file:") :].strip()
+            version = self._read_version_from_file(file_arg)
+        elif version.startswith("attr:"):
+            attr_arg = version[len("attr:") :].strip()
```
This is the CI failure (`black` and `flake8` don't agree on the space before `:`) - permission to silence flake8?

MartinBasti replied:
why space before `:` ?

chmeliik replied:
IDK, that's how `black` wants to format it

MartinBasti replied:
ok, I disagree with black

chmeliik replied:
Silenced `black` instead, if anyone disagrees let me know :-)

lcarva replied:
Can we adjust the configuration instead?

chmeliik replied:
`black` doesn't have any configuration, from the [README](https://github.com/psf/black/blob/master/README.md):

> Black is a PEP 8 compliant opinionated formatter. Black reformats entire files in place. It is not configurable. It doesn't take previous formatting into account. Your main option of configuring Black is that it doesn't reformat blocks that start with # fmt: off and end with # fmt: on.

We could adjust the `flake8` config, not sure what would be the preference

lcarva replied:
The rule of thumb on this project so far has been to always agree with `black`. Even though we may disagree with some of its choices, it's still worth for all the other parts that it gets correct. I suggest we disable the `E203` check in flake8.

chmeliik replied:
Ok, I'll go with `black`. E203 seems valid outside of this specific scenario, I'll just disable the `flake8` check inline

chmeliik replied:
Nevermind, `black` will catch the valid cases, disabling in config

chmeliik commented:
Might want to add logging checks to unit tests, other than that should be pretty much done

chmeliik commented:
Removed duplicate test case

athos-ribeiro commented on docker/Dockerfile-api:
```diff
@@ -21,6 +21,7 @@ RUN dnf -y install \
     python3-requests \
     python3-requests-kerberos \
     python3-semver \
+    python3-setuptools \
```
Do we really need this here as well?

chmeliik commented:
Added logging checks and finally figured out why they were not working

chmeliik commented on docker/Dockerfile-api:
```diff
@@ -21,6 +21,7 @@ RUN dnf -y install \
     python3-requests \
     python3-requests-kerberos \
     python3-semver \
+    python3-setuptools \
```
Not sure, probably not. As far as I can tell, `Dockerfile-api` installs everything in `requirements.txt + requirements-web.txt`, and `Dockerfile-workers` installs only `requirements.txt`, so I wanted to keep it that way

athos-ribeiro replied:
Fair enough

chmeliik commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+        if hasattr(version, "__iter__"):
+            version = ".".join(map(str, version))
+        else:
+            version = str(version)
```
Added, thanks!

chmeliik commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+        if "\n" in package_dir_value:
+            package_items = package_dir_value.splitlines()
+        else:
+            package_items = package_dir_value.split(",")
```
Added, thanks!

athos-ribeiro commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+
+        :rtype: str or None
+        """
+        self._parse()
```
Why are we calling `self._parse()` both here and in `get_version` and using `self._parsed` in `get_option` if `self._parse()` sets and returns `self._parsed`? i.e., why don't we use `self._parse()` in `get_option` and remove the other self._parse() calls?

Alternatively, we could remove the `self._parse()` return value since it is not being used.

lcarva replied:
Conversely, change `_parsed` to `__parsed`, and create a property `_parsed`. That way it gets parsed when, and only when, it's needed.

chmeliik replied:
That is better, thanks! I also realized I'm currently ignoring the possibility that setup.cfg parsing would fail, not sure if that should be `ValidationError` or a "failed to resolve", I'm thinking the latter

lcarva commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+            package_items = package_dir_value.split(",")
+
+        # Strip whitespace and discard empty values
+        package_items = filter(bool, (p.strip() for p in package_items))
```
This is neat! It avoids the repetition seen in a pure list comprehension approach: 
```python
(p.strip() for p in package_items if p.strip())
```

chmeliik replied:
Right? That's pretty much exactly what setuptools does in their code, so of course I thought: "Ha! I can do you one better."

chmeliik commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+    """
+
+    # Valid Python name - any sequence of \w characters that does not start with a number
+    _name_re = re.compile(r"[^\W\d]\w*")
```
FWIW, the pattern does also support non-ASCII names:
```python
In [7]: pat = re.compile(r"[^\W\d]\w*")                                                       

In [8]: pat.fullmatch('Žluťoučký')                                                            
Out[8]: <re.Match object; span=(0, 9), match='Žluťoučký'>

In [9]: pat.fullmatch('こんにちは')                                                           
Out[9]: <re.Match object; span=(0, 5), match='こんにちは'>
```
So we might get false positives (at least on py2), but unless the user is actively trying to fail their build, it should be fine :D

chmeliik commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+        self._parse()
+        name = self._get_option("metadata", "name")
+        if name is None:
+            log.debug("No metadata.name in setup.cfg")
```
I suppose I could use `self._path.name` instead, but that will just end up being more typing with not that much upside. I was thinking of somehow injecting the filename into the log messages automatically, but I don't have much experience with more advanced Python logging. Ideally, it would be something like
```python
self.log.debug("Found something in %(filename)s: %s", some_value)  # automatically injects the `filename` arg
```
But so far, I haven't figured out how to do that properly, any ideas?

Worst case scenario, I can fall back to using `self._path.name` everywhere if desirable

chmeliik commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,321 @@
+        Initialize a SetupFile.
+
+        :param str top_dir: Path to root of project directory
+        :param str file_name: Either "setup.cfg" or "setup.py"
```
Good point, changed

chmeliik commented:
Added handling for malformed `.cfg` files and the corresponding unit test scenarios

athos-ribeiro commented on tox.ini:
```diff
@@ -42,7 +42,8 @@ show-source = True
 # W503 line break before binary operator
-ignore = D100,D104,D105,W503
+# E203 whitespace before ':'
+ignore = D100,D104,D105,W503,E203
```
The reason this is not disabled inline, is that black will also deal with the same issue. Is this right? If so, can we add a comment on that?

chmeliik replied:
Exactly, added comment

chmeliik commented on cachito/workers/pkg_managers/pip.py:
```diff
@@ -0,0 +1,320 @@
+        """
+        module_name, _, attr_name = attr_spec.rpartition(".")
+        if not module_name:
+            # Assume current directory is a module, look for attribute in __init__.py
```
Yes it should! Thanks :-)

lcarva commented on tests/test_workers/test_pkg_managers/test_pip.py:
```diff
@@ -0,0 +1,587 @@
+            (
+                dedent(
+                    """
+            [metadata]
```
Is Black pushing for thins indentation? If so, try adding a backslash after `"""`:
```python
dedent(
  """\
  [metadata]
  ...
  """
)

chmeliik replied:
Oh, that feels *so* much better :D

lcarva commented on tests/test_workers/test_pkg_managers/test_pip.py:
```diff
@@ -0,0 +1,587 @@
+    try:
+        yield
+    finally:
+        remove_file_tree(tree_def, rooted_at)
```
Why not just `shutil.rmtree`? Or, better yet, always use `tmpdir` and let pytest clean up these files automatically.

chmeliik replied:
Good point, I had some issues with leftover files when I was trying to use `importlib.util.find_spec()` for finding modules (which also involved temporarily modifying `sys.path`). In the end, I never got it to work and the current solution no longer has this issue.

chmeliik commented on tests/test_workers/test_pkg_managers/test_pip.py:
```diff
@@ -0,0 +1,587 @@
+                },
+                "1.0.0",
+                None,
+                [],
```
That's fair, I'll make the logging checks as complete as possible

chmeliik replied:
Added

chmeliik commented:
Added missing checks for log messages. Moved the function which actually runs the tests from the bottom of the file to where it makes more sense.

No clue why CI is failing, pytest says
```
  File "/home/travis/build/release-engineering/cachito/.tox/py37/lib/python3.7/site-packages/celery/utils/log.py", line 13, in <module>
    from kombu.five import PY3, values
ModuleNotFoundError: No module named 'kombu.five'
```

chmeliik commented:
Figured out what was wrong with CI, see 2nd commit:
```
Celery requires kombu>=4.6.10,<4.7; this is not respected during
installation (probably because we specify 'kombu' explicitly).
```
 Not sure if that is the right way to fix it though, is there any better way to get around the issue?

athos-ribeiro commented:
> Figured out what was wrong with CI, see 2nd commit:
> 
> ```
> Celery requires kombu>=4.6.10,<4.7; this is not respected during
> installation (probably because we specify 'kombu' explicitly).
> ```
> 
> Not sure if that is the right way to fix it though, is there any better way to get around the issue?

Note that kombu 5.0.0 was released just a few hours ago, but celery 5 is still in alpha.

For now, we user Celery from Fedora 31 in our container images, which is python-celery-4.3.0-2.fc31. This relies con komub < 5, so we can just set the requirements versions as so. (both celery and kombu should be set to be < 5). 

Then we should file a ticket to support Celery 5 later, whenever it lands in Fedora.

WDYT?

chmeliik commented:
> Note that kombu 5.0.0 was released just a few hours ago, but celery 5 is still in alpha.
> 
> For now, we use Celery from Fedora 31 in our container images, which is python-celery-4.3.0-2.fc31. This relies on komub < 5, so we can just set the requirements versions as so. (both celery and kombu should be set to be < 5).
> 
> Then we should file a ticket to support Celery 5 later, whenever it lands in Fedora.
> 
> WDYT?

Sounds good to me, I'll make a separate PR for the requirements to unblock other people's PRs (someone else will have to merge it though)

athos-ribeiro commented:
Nice! Could you remove the workaround commit now?
